---
output: github_document
editor_options: 
  chunk_output_type: console
---
<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r echo = FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

# Apjok-et-al-DEEPMINE-NatMicrobiol

Code deposited for paper "Expanding the host-range of functional metagenomics reveals resistance threats to novel antibiotics", Apjok et al. Nature Microbiology.

## System requirements

* [Apptainer/Singularity](https://apptainer.org/) (for demultiplexing, only works on linux)
* [R](https://www.r-project.org/) (for all other scripts)

## Download data

Illumina reads and nanopore contigs can be downloaded from the European Nucleotide Archive. Study Accession: PRJEB54063 (Secondary Accession: ERP138883)

```{r echo = FALSE, eval = FALSE}
suppressMessages(source("prepare_input_data.R"))
```

## Demultiplex illumina reads

Download Apptainer/Singularity container for demultiplexing:

```{r}
# Define a local directory that will store the apptainer/singularity container
cache_dir <- "./containers"

# convert path to absolute path
cache_dir <- normalizePath(cache_dir)

# create cache directory if it doesn't already exists
if (!dir.exists(cache_dir)) {
  dir.create(cache_dir, recursive = TRUE)
}

# download container image if it is not already downloaded
if (!file.exists(paste0(cache_dir, "/mesti90-hgttree-latest.img"))) {
  cmd <- paste0(
    "singularity pull --dir ", cache_dir,
    " mesti90-hgttree-latest.img docker://mesti90/hgttree:latest"
  )
  system(cmd)
}
```

Demultiplex Illumina reads in first run:

```{r}
setwd("./run1")
cmd <- paste0(
  "singularity exec ", cache_dir,"/mesti90-hgttree-latest.img ",
  "python3 barcode_amplicon_demux_20201026.py3"
)
system(cmd, ignore.stdout = TRUE)
```

Demultiplex illumina reads in second run:

```{r}
setwd("./run2")
cmd <- paste0(
  "singularity exec ", cache_dir,"/mesti90-hgttree-latest.img ",
  "python3 barcode_amplicon_demux_20210201.py3"
)
system(cmd, ignore.stdout = TRUE)
```

Demultiplex illumina reads in third run:

```{r}
setwd("./run3")
cmd <- paste0(
  "singularity exec ", cache_dir,"/mesti90-hgttree-latest.img ",
  "python3 barcode_amplicon_demux_20220505.py3"
)
system(cmd, ignore.stdout = TRUE)
```

## Prepare supplementary table 6

```{r}
setwd("./supplementary_table_6")
suppressMessages(source("Script - Supplementary table 6.R"))
```

## Prepare supplementary table 7

```{r}
setwd("./supplementary_table_7")
suppressMessages(source("Script - Supplementary table 7.R"))
```

## Prepare supplementary table 9

```{r}
setwd("./supplementary_table_9")
suppressMessages(source("Script - Supplementary table 9.R"))
```
